### ISSUE TITLE
Can't find unit Graphics used by IntfGraphics (Lazarus Upgrade)

### ISSUE THREAD

**Issue by BrainWaveCC** [LongDirtyAnimAlf/fpcupdeluxe#504]
I am using version 2.2.0i of FPCupDeluxe, and I cannot get any version of Lazarus to install.

I have gotten FPC 3.2.0, 3.2.2 and fixes-3.2 to install successful (by themselves)
But, trying to install Lazarus fixes failed.
-- so did 2.2.0
-- so did 2.3.0
-- so did fixes-2.0
-- so did fixes-2.2

The Lazarus installs all fail with the following:

fpcupdeluxe: info: LazarusNativeInstaller (GetModule: Lazarus): Lazarus was at revision/hash: lazarus_2_2_0-0-g4d49533f10
fpcupdeluxe: info: LazarusNativeInstaller (GetModule: Lazarus): Lazarus is now at revision/hash: lazarus_2_2_0-64-gff329cf453
fpcupdeluxe: info: Lazarus native builder: Detected source version Lazarus: 2.2.0
fpcupdeluxe: info: Lazarus native builder: Using FPC compiler with version: 3.2.3
fpcupdeluxe: Start of compile error summary.
Fatal: (10022) Can't find unit Graphics used by IntfGraphics
(10028) Recompiling Graphics, checksum changed for C:\Programs\Lazarus.Fixes\lazarus\components\lazutils\lib\x86_64-win64\graphmath.ppu
fpcupdeluxe: Start of compile error summary.
Fatal: (1018) Compilation aborted
Fatal: (10022) Can't find unit Graphics used by IntfGraphics
fpcupdeluxe: Start of compile error summary.
Error: C:\Programs\Lazarus.Fixes\fpc\bin\x86_64-win64\ppcx64.exe returned an error exitcode
fpcupdeluxe: Start of compile error summary.
fpcupdeluxe: ERROR: LazarusNativeInstaller (BuildModuleCustom: Lazbuild): make.exe returned exit status #2
fpcupdeluxe: Done !!

I renamed the PPU file mentioned above, and tried it again, and received the very same error.

![image](https://user-images.githubusercontent.com/48736709/152846630-34ba0d2f-6e82-4876-a61c-840f9445dfae.png)

**Comment by BrainWaveCC** [comment]
Tried to upgrade my FPC/Lazarus install on another system, and received a very similar error message:

fpcupdeluxe: info: Lazarus native builder: Detected source version Lazarus: 2.2.0
fpcupdeluxe: info: Lazarus native builder: Using FPC compiler with version: 3.2.3
fpcupdeluxe: Start of compile error summary.
Fatal: (10022) Can't find unit Graphics used by IntfGraphics
(10028) Recompiling Graphics, checksum changed for C:\Programs\Lazarus.Fixes\lazarus\components\lazutils\lib\x86_64-win64\graphmath.ppu
fpcupdeluxe: Start of compile error summary.
Fatal: (1018) Compilation aborted
Fatal: (10022) Can't find unit Graphics used by IntfGraphics
fpcupdeluxe: Start of compile error summary.
Error: C:\Programs\Lazarus.Fixes\fpc\bin\x86_64-win64\ppcx64.exe returned an error exitcode
fpcupdeluxe: Start of compile error summary.
fpcupdeluxe: ERROR: LazarusNativeInstaller (BuildModuleCustom: Lazbuild): make.exe returned exit status #2
fpcupdeluxe: Done !!

![image](https://user-images.githubusercontent.com/48736709/152874364-da51c453-bc77-4266-9543-d20a73ca55ab.png)

The first system I did this on was running Windows 11, btw.  But this second system is running Windows 10 (like all my other systems).

**Comment by BrainWaveCC** [comment]
I need to clarify that these were not clean installs, but were upgrades over earlier FPCupDeluxe installs of earlier, functional versions of FPC/Lazarus.

The previous x64 installation contained the following:
-- FreePascal 3.2.1 (fixes-r49055) + Lazarus 2.0.13 (fixes-r64843)

With Windows x86, and Linux/Arm cross-compiles.

I just tried a new install to a different folder, with the same config as above, and it did work.  I had to copy back some of my configuration files (I ultimately renamed the folders of the new install and old install to make the migration easier).

I'm going to verify this with one more of my machines, and then I'll likely close the ticket if it works, but I would love to know why it failed the first time, and if there was a different way around it.

**Comment by resistancelion** [comment]
> I need to clarify that these were not clean installs, but were upgrades over earlier FPCupDeluxe installs of earlier, functional versions of FPC/Lazarus.
> 
> The previous x64 installation contained the following: -- FreePascal 3.2.1 (fixes-r49055) + Lazarus 2.0.13 (fixes-r64843)
> 
> With Windows x86, and Linux/Arm cross-compiles.
> 
> I just tried a new install to a different folder, with the same config as above, and it did work. I had to copy back some of my configuration files (I ultimately renamed the folders of the new install and old install to make the migration easier).
> 
> I'm going to verify this with one more of my machines, and then I'll likely close the ticket if it works, but I would love to know why it failed the first time, and if there was a different way around it.

Uninstalling cross-compilers and then upgrading also works

**Comment by LongDirtyAnimAlf** [comment]
Thanks for trying to pinpoint this issue !
It is definitely caused by another install that on the same machine.
However, this should NOT happen, so I am still in the dark about this issue.

**Comment by BrainWaveCC** [comment]
> Uninstalling cross-compilers and then upgrading also works

This did not work for me on my machine.

Right now, I have three systems running FPC/Lazarus.

I have one system (LAPTOP) which works, but where I did the following:
- Renamed folder A to A.old
- Installed using v2.2.0i to folder A
- Copied a few config files over to A from A.old, once completed.
- Completely functional
**- Follow-up: Just tried another upgrade and it failed.**


I have one system (this primary one, DESKTOP) where I am still using the older build (pre Git)

And I have the test system (DESKTOP) where I have tried a variety of things including:
- Upgrade both FPC and Lazarus (failed)
- Upgrade FPC only (succeed)
- Upgrade Lazarus only (failed)
- Upgrade both cross-compilers (succeed)
- Upgrade Lazarus only (failed)
- Uninstall both cross-compilers (succeed)
- Upgrade Lazarus only (failed, with new error)

I will play with this for a few more days -- maybe a week -- and then I will just use the method I did on the first machine.  And if it works (which I fully expect), then I will do it on my primary machine also.


**(Follow-up)**
I just tried to upgrade the first system listed above, and it also failed.

![image](https://user-images.githubusercontent.com/48736709/154610843-fa5ef463-7108-4da5-b949-0060193f6aca.png)

**Comment by LongDirtyAnimAlf** [comment]
Thanks for your update.
However. The errors reported are different from each other.
I would prefer to limit this issue to its title and try to solve that issue.

**Comment by BrainWaveCC** [comment]
> Thanks for your update.
> However. The errors reported are different from each other.
> I would prefer to limit this issue to its title and try to solve that issue.

Fair enough, but what do you propose as the next step.

While I know how to get around the problem one time, what I am now seeing is that even after the first fix, I am unable to upgrade any longer...

**Comment by BrainWaveCC** [comment]
> Uninstalling cross-compilers and then upgrading also works

That hasn't been as helpful for me.

I have, however, seen better outcomes by doing the following:

- Upgrade FPC only
- Upgrade cross-compilers (one at a time or in bulk doesn't matter)
- Upgrade FPC only (again)
- NOW try to upgrade Lazarus (if this errors, do FPC only, then Lazarus only)

Slow, but has worked twice now.   I'm about to try it with the instance that has been untouched so far.

**Comment by BrainWaveCC** [comment]
> Thanks for your update.
> However. The errors reported are different from each other.
> I would prefer to limit this issue to its title and try to solve that issue.

I finally tried to tackle the FPC/Lazarus install on my last (primary) system, and this is what I get all day for the Lazarus upgrade portion...

-------
.
.
.
(1008) 7699 lines compiled, 0.5 sec
C:/Programs/Lazarus.Fixes/fpcbootstrap/make.exe -C components/lazcontrols LCL_PLATFORM=nogui
C:/Programs/Lazarus.Fixes/fpc/bin/x86_64-win64/fpc.exe -MObjFPC -Scghi -O1 -g -gl -l -vewnhibq -dLCL -dLCLnogui -Fu../../packager/units/x86_64-win64 -Fu../lazutils/lib/x86_64-win64 -Fu../freetype/lib/x86_64-win64 -Fu../../lcl/units/x86_64-win64 -Fu../../lcl/units/x86_64-win64/nogui -Fu. -FuC:/Programs/Lazarus.Fixes/fpcsrc/rtl -FE. -FUlib/x86_64-win64/nogui -vw-n-h-l-d-u-t-p-c- -g -gl -dx86_64 lazcontrols.pas
(1002) Target OS: Win64 for x64
(3104) Compiling lazcontrols.pas
(3104) Compiling checkboxthemed.pas
(10001) PPU Loading C:\Programs\Lazarus.Fixes\lazarus\lcl\units\x86_64-win64\graphics.ppu
(10011) PPU Source: graphics.pp not found
(10011) PPU Source: lcl_defines.inc not available
(10011) PPU Source: graphicsobject.inc not available
(10011) PPU Source: graphic.inc not available
(10011) PPU Source: picture.inc not available
(10011) PPU Source: sharedimage.inc not available
(10011) PPU Source: sharedrasterimage.inc not available
(10011) PPU Source: sharedcustombitmap.inc not available
(10011) PPU Source: rasterimage.inc not available
(10011) PPU Source: custombitmap.inc not available
(10011) PPU Source: bitmapcanvas.inc not available
(10011) PPU Source: pen.inc not available
(10011) PPU Source: brush.inc not available
(10011) PPU Source: region.inc not available
(10011) PPU Source: font.inc not available
(10011) PPU Source: canvas.inc not available
(10011) PPU Source: pixmap.inc not available
(10011) PPU Source: png.inc not available
(10011) PPU Source: pnm.inc not available
(10011) PPU Source: jpegimage.inc not available
(10011) PPU Source: icon.inc not available
(10011) PPU Source: icnsicon.inc not available
(10011) PPU Source: cursorimage.inc not available
(10011) PPU Source: fpimagebitmap.inc not available
(10011) PPU Source: bitmap.inc not available
(10011) PPU Source: tiffimage.inc not available
(10011) PPU Source: gifimage.inc not available
(10011) PPU Source: patternbitmap.inc not available
(10001) PPU Loading C:\Programs\Lazarus.Fixes\lazarus\lcl\units\x86_64-win64\graphics.ppu
(10011) PPU Source: graphics.pp not found
(10011) PPU Source: lcl_defines.inc not available
(10011) PPU Source: graphicsobject.inc not available
(10011) PPU Source: graphic.inc not available
(10011) PPU Source: picture.inc not available
(10011) PPU Source: sharedimage.inc not available
(10011) PPU Source: sharedrasterimage.inc not available
(10011) PPU Source: sharedcustombitmap.inc not available
(10011) PPU Source: rasterimage.inc not available
(10011) PPU Source: custombitmap.inc not available
(10011) PPU Source: bitmapcanvas.inc not available
(10011) PPU Source: pen.inc not available
(10011) PPU Source: brush.inc not available
(10011) PPU Source: region.inc not available
(10011) PPU Source: font.inc not available
(10011) PPU Source: canvas.inc not available
(10011) PPU Source: pixmap.inc not available
(10011) PPU Source: png.inc not available
(10011) PPU Source: pnm.inc not available
(10011) PPU Source: jpegimage.inc not available
(10011) PPU Source: icon.inc not available
(10011) PPU Source: icnsicon.inc not available
(10011) PPU Source: cursorimage.inc not available
(10011) PPU Source: fpimagebitmap.inc not available
(10011) PPU Source: bitmap.inc not available
(10011) PPU Source: tiffimage.inc not available
(10011) PPU Source: gifimage.inc not available
(10011) PPU Source: patternbitmap.inc not available
(10028) Recompiling Graphics, checksum changed for C:\Programs\Lazarus.Fixes\lazarus\components\lazutils\lib\x86_64-win64\graphmath.ppu
Fatal: (10022) Can't find unit Graphics used by IntfGraphics
Fatal: (1018) Compilation aborted
Error: C:\Programs\Lazarus.Fixes\fpc\bin\x86_64-win64\ppcx64.exe returned an error exitcode
C:\Programs\Lazarus.Fixes\fpcbootstrap\make.exe: *** [lazbuild] Error 2
fpcupdeluxe: ERROR: LazarusNativeInstaller (BuildModuleCustom: Lazbuild): make.exe returned exit status #2


ERROR: Fpcupdeluxe fatal error !
Sequencer (UserIDE): Failure running fpcupdeluxe: error executing sequence UserIDE
Sequencer (Lazarus): Failure running fpcupdeluxe: error executing sequence Lazarus
Sequencer (Only): Failure running fpcupdeluxe: error executing sequence Only

![image](https://user-images.githubusercontent.com/48736709/155041201-46b43cfa-9363-4c61-9be2-c27dc2aea40a.png)

**Comment by BrainWaveCC** [comment]
Still having this issue...

Everything is going well until:

C:/Programs/Lazarus.Fixes/fpc/bin/x86_64-win64/fpc.exe -MObjFPC -Scghi -O1 -g -gl -l -vewnhibq -dLCL -dLCLnogui -Fu../../packager/units/x86_64-win64 -Fu../lazutils/lib/x86_64-win64 -Fu../freetype/lib/x86_64-win64 -Fu../../lcl/units/x86_64-win64 -Fu../../lcl/units/x86_64-win64/nogui -Fu. -FuC:/Programs/Lazarus.Fixes/fpcsrc/rtl -FE. -FUlib/x86_64-win64/nogui -vw-n-h-l-d-u-t-p-c- -g -gl -dx86_64 lazcontrols.pas
(1002) Target OS: Win64 for x64
(3104) Compiling lazcontrols.pas
(3104) Compiling checkboxthemed.pas
(10001) PPU Loading C:\Programs\Lazarus.Fixes\lazarus\lcl\units\x86_64-win64\graphics.ppu
(10011) PPU Source: graphics.pp not found
(10011) PPU Source: lcl_defines.inc not available
(10011) PPU Source: graphicsobject.inc not available
(10011) PPU Source: graphic.inc not available
(10011) PPU Source: picture.inc not available
(10011) PPU Source: sharedimage.inc not available
(10011) PPU Source: sharedrasterimage.inc not available
(10011) PPU Source: sharedcustombitmap.inc not available
(10011) PPU Source: rasterimage.inc not available
(10011) PPU Source: custombitmap.inc not available
(10011) PPU Source: bitmapcanvas.inc not available
(10011) PPU Source: pen.inc not available
(10011) PPU Source: brush.inc not available
(10011) PPU Source: region.inc not available
(10011) PPU Source: font.inc not available
(10011) PPU Source: canvas.inc not available
(10011) PPU Source: pixmap.inc not available
(10011) PPU Source: png.inc not available
(10011) PPU Source: pnm.inc not available
(10011) PPU Source: jpegimage.inc not available
(10011) PPU Source: icon.inc not available
(10011) PPU Source: icnsicon.inc not available
(10011) PPU Source: cursorimage.inc not available
(10011) PPU Source: fpimagebitmap.inc not available
(10011) PPU Source: bitmap.inc not available
(10011) PPU Source: tiffimage.inc not available
(10011) PPU Source: gifimage.inc not available
(10011) PPU Source: patternbitmap.inc not available
(10001) PPU Loading C:\Programs\Lazarus.Fixes\lazarus\lcl\units\x86_64-win64\graphics.ppu
(10011) PPU Source: graphics.pp not found
(10011) PPU Source: lcl_defines.inc not available
(10011) PPU Source: graphicsobject.inc not available
(10011) PPU Source: graphic.inc not available
(10011) PPU Source: picture.inc not available
(10011) PPU Source: sharedimage.inc not available
(10011) PPU Source: sharedrasterimage.inc not available
(10011) PPU Source: sharedcustombitmap.inc not available
(10011) PPU Source: rasterimage.inc not available
(10011) PPU Source: custombitmap.inc not available
(10011) PPU Source: bitmapcanvas.inc not available
(10011) PPU Source: pen.inc not available
(10011) PPU Source: brush.inc not available
(10011) PPU Source: region.inc not available
(10011) PPU Source: font.inc not available
(10011) PPU Source: canvas.inc not available
(10011) PPU Source: pixmap.inc not available
(10011) PPU Source: png.inc not available
(10011) PPU Source: pnm.inc not available
(10011) PPU Source: jpegimage.inc not available
(10011) PPU Source: icon.inc not available
(10011) PPU Source: icnsicon.inc not available
(10011) PPU Source: cursorimage.inc not available
(10011) PPU Source: fpimagebitmap.inc not available
(10011) PPU Source: bitmap.inc not available
(10011) PPU Source: tiffimage.inc not available
(10011) PPU Source: gifimage.inc not available
(10011) PPU Source: patternbitmap.inc not available
(10028) Recompiling Graphics, checksum changed for C:\Programs\Lazarus.Fixes\lazarus\components\lazutils\lib\x86_64-win64\graphmath.ppu
Fatal: (10022) Can't find unit Graphics used by IntfGraphics
Fatal: (1018) Compilation aborted
Error: C:\Programs\Lazarus.Fixes\fpc\bin\x86_64-win64\ppcx64.exe returned an error exitcode
C:\Programs\Lazarus.Fixes\fpcbootstrap\make.exe: *** [lazbuild] Error 2
fpcupdeluxe: ERROR: LazarusNativeInstaller (BuildModuleCustom: Lazbuild): make.exe returned exit status #2


ERROR: Fpcupdeluxe fatal error !
Sequencer (UserIDE): Failure running fpcupdeluxe: error executing sequence UserIDE
Sequencer (Lazarus): Failure running fpcupdeluxe: error executing sequence Lazarus
Sequencer (Default): Failure running fpcupdeluxe: error executing sequence Default

**Comment by BrainWaveCC** [comment]
I ran the following command manually, and it generated errors until I did two things:

C:\Programs\Lazarus.Fixes\fpc\bin\x86_64-win64\fpc.exe -MObjFPC -Scghi -O1 -g -gl -l -vewnhibq -dLCL -dLCLnogui -Fu..\..\packager\units\x86_64-win64 -Fu..\lazutils\lib\x86_64-win64 -Fu..\freetype\lib\x86_64-win64 -Fu..\..\lcl\units\x86_64-win64 -Fu..\..\lcl\units\x86_64-win64\nogui -Fu. -FuC:\Programs\Lazarus.Fixes\fpcsrc\rtl -FE. -FUlib\x86_64-win64\nogui -vw-n-h-l-d-u-t-p-c- -g -gl -dx86_64 lazcontrols.pas

1. I started the command while in the \lazarus\components\lazcontrols folder
2. I added the following parameters:  **-Fu**C:\Programs\Lazarus.Fixes\\**lazarus\lcl** **-Fu**C:\Programs\Lazarus.Fixes\\**lazarus\lcl\include**

**Comment by BrainWaveCC** [comment]
> Thanks for your update. However. The errors reported are different from each other. I would prefer to limit this issue to its title and try to solve that issue.

I believe I have found the issue.  Not sure why it happens in an upgrade scenario, but not a new install into a clean folder.

**Comment by BrainWaveCC** [comment]
Here is the output at the command point, with both the old command and the new command...
[2022-03-02 FpcUpDeluxe.TXT](https://github.com/LongDirtyAnimAlf/fpcupdeluxe/files/8172778/2022-03-02.FpcUpDeluxe.TXT)

**Comment by BrainWaveCC** [comment]
Hey, @LongDirtyAnimAlf, have you had any time to take a peek at the last few posts?

**Comment by LongDirtyAnimAlf** [comment]
No. Sorry.
I have problems with my back.
And some business related things to handle.
So, I lack time and motivation unfortunately.
However, this issue will be handled naturally !
Please permit me a bit more time (to heal).

**Comment by BrainWaveCC** [comment]
Hi Don, 

Sorry to hear about your current situation.  Please get well soon.

**Comment by LongDirtyAnimAlf** [comment]
I am back.
Please update me about your latest findings please.
Did you pinpoint the error yourself ?
And can you direct me to the error once more (it has been a long time away for me) ?
Thanks !

**Comment by BrainWaveCC** [comment]
Hi Don,

This link outlines where I found the error:
https://github.com/LongDirtyAnimAlf/fpcupdeluxe/issues/504#issuecomment-1057353038

For reasons I cannot decipher, it seems like it ignores key components in an upgrade vs a new install.

**Comment by BrainWaveCC** [comment]
Hi Don, 

I tried to use 2.2.0j-prerelease, and I am still getting the same issue if I try to install both FPC/Lazarus at the same time.

FPC updates to FIXES level okay by itself.
Lazarus gives a different error if I try to install 2.3.0 by itself.

> graphmath.pp(744,13) Error: Identifier not found "TruncToInt"
> graphmath.pp(1053) Fatal: There were 1 errors compiling module, stopping
> Fatal: Compilation aborted
> Error: C:\Programs\Lazarus.Fixes\fpc\bin\x86_64-win64\ppcx64.exe returned an error exitcode
> C:\Programs\Lazarus.Fixes\fpcbootstrap\make.exe: *** [lazbuild] Error 2
> fpcupdeluxe: ERROR: LazarusNativeInstaller (BuildModuleCustom: Lazbuild): make.exe returned exit status #2
> 
> 
> ERROR: Fpcupdeluxe fatal error !
> Sequencer (UserIDE): Failure running fpcupdeluxe: error executing sequence UserIDE
> Sequencer (Lazarus): Failure running fpcupdeluxe: error executing sequence Lazarus
> Sequencer (Only): Failure running fpcupdeluxe: error executing sequence Only
> 

Lazarus gives the same error as before if I try to install 2.2.2 by itself.

> Fatal: (10022) Can't find unit Graphics used by IntfGraphics
> Fatal: (1018) Compilation aborted
> Error: C:\Programs\Lazarus.Fixes\fpc\bin\x86_64-win64\ppcx64.exe returned an error exitcode
> C:\Programs\Lazarus.Fixes\fpcbootstrap\make.exe: *** [lazbuild] Error 2
> fpcupdeluxe: ERROR: LazarusNativeInstaller (BuildModuleCustom: Lazbuild): make.exe returned exit status #2
> 
> 
> ERROR: Fpcupdeluxe fatal error !
> Sequencer (UserIDE): Failure running fpcupdeluxe: error executing sequence UserIDE
> Sequencer (Lazarus): Failure running fpcupdeluxe: error executing sequence Lazarus
> Sequencer (Only): Failure running fpcupdeluxe: error executing sequence Only
>

**Comment by LongDirtyAnimAlf** [comment]
I cannot reproduce unfortunately.
If I do a vanilla FPC + Lazarus stable install, all is well.

**Comment by BrainWaveCC** [comment]
> If I do a vanilla FPC + Lazarus stable install, all is well.

Yes, I was able to do that when testing as well.  But, once I add back the components I was using from before, the next upgrade doesn't work again.

I'll verify that this still happens with one of my installs.  (I have to keep at least one of them working, so I only have two to test with.)

Is there a place where I can upload to you the install folder (compressed) as it now is, to aid in the troubleshooting?

**Comment by LongDirtyAnimAlf** [comment]
I am closing this issue due to the fact that its not something that can be solved by fpcupdeluxe.

**Comment by BrainWaveCC** [comment]
After a break to deal with some other things, I put some more effort into testing this, and I started making progress when I searched for each problematic .pas file and deleted it.

I don't know why I had so many historical files in there, but I finally deleted all .pas files in the destination, prior to Oct 2022, and then ran the install again.  It finally worked flawlessly.

Not sure what made it break all of a sudden back in Feb, when I've been updating it steadily for a couple years now, but getting rid of all the old .pas files in the destination appears to have totally resolved the issue for me.
